dali/operators/generic/histogram/histogram.cc

```diff
@@ -0,0 +1,685 @@
+  norm_outputs = collapse_dims(norm_outputs, out_collapse_groups);
+
+  auto splited_input_shapes = unfold_outer_dim(norm_inputs);
+  auto splited_output_shapes = unfold_outer_dim(norm_outputs);
+
+  std::vector<int> split_mapping;
+  split_mapping.reserve(splited_input_shapes.num_samples());
+  auto norm_non_reduced = norm_inputs.first(1);
+  for (int i = 0; i < norm_non_reduced.num_samples(); ++i) {
+    for (int j = 0; j < norm_non_reduced[i][0]; ++j) {
+      split_mapping.push_back(i);
+    }
+  }
+
+  split_mapping_ = std::move(split_mapping);
+  splited_input_shapes_ = std::move(splited_input_shapes);
+  splited_output_shapes_ = std::move(splited_output_shapes);
+
+  output_desc.shape = std::move(output_shapes);
+  output_desc.type = DALI_FLOAT;
+}
+
+bool HistogramCPU::SetupImpl(std::vector<OutputDesc> &output_desc,
+                             const workspace_t<CPUBackend> &ws) {
+  output_desc.resize(1);
+
+  auto &input = ws.template Input<CPUBackend>(0);
+  const size_t ndims = input.shape().sample_dim();
+
+  hist_dim_ = ValidateRangeArguments(ws, input.num_samples());
+
+  PrepareReductionAxes(input.GetLayout(), ndims, hist_dim_);
+  ValidateInputShape(input.shape(), hist_dim_, channel_axis_);
+
+  // If an empty reduction axes were specified, histogram calculation becomes identity operation
+  if (is_identity_) {
+    output_desc[0].type = input.type();
+    output_desc[0].shape = input.shape();
+  } else {
+    PrepareReductionShapes(input.shape(), output_desc[0]);
+  }
+
+  return true;
+}
+
+void HistogramCPU::ValidateBinsArgument(const workspace_t<CPUBackend> &ws, int num_samples, int hist_dim) {
+  assert(uniform_ && "Specified only for uniform histogram");
+  param_num_bins_.Acquire(this->spec_, ws, num_samples, TensorShape<1>(hist_dim));
+  auto bins_view = param_num_bins_.get();
+
+  // This always passes meaning that num_bins is broadcasted when it shouldn't for ND-histogram
+  DALI_ENFORCE(bins_view.shape.num_elements()/num_samples == hist_dim);
+
+  DALI_ENFORCE(bins_view.num_elements() % num_samples == 0,
+               make_string("Histogram bins should be an array of bins per sample",
+                           bins_view.num_elements()));
+
+  batch_bins_.resize(0);
+  batch_bins_.reserve(num_samples);
+  for (int i=0; i<num_samples; ++i) {
+    SmallVector<int, 3> bins;
+    for (int j=0; j<hist_dim; ++j) {
+      bins.push_back(bins_view.tensor_data(i)[j]);
+    }
+    batch_bins_.push_back(std::move(bins));
+  }
+}
+
+void HistogramCPU::InferBinsArgument(const workspace_t<CPUBackend> &ws, int num_samples, int hist_dim) {
+  assert(!uniform_ && "Can infer only for non-uniform histograms");
+
+  SmallVector<int, 3> bins;
+  for (int i=0; i < hist_dim; ++i) {
+    const auto &dim_ranges = ws.template Input<CPUBackend>(1 + i);
+    bins.push_back(dim_ranges.shape().num_samples());
+  }
+  batch_bins_.reserve(num_samples);
+
+  for (int i=0; i<num_samples; ++i) {
+    batch_bins_.push_back(bins);
+  }
+}
```

Comment:
I don't understand what this is trying to accomplish. Why would `num_samples` (number of samples in a batch) be related to bin sizes?